TTS/tts/models/naturalspeech2.py

```diff
@@ -0,0 +1,1223 @@
+            x (torch.FloatTensor): Input sequence.
+            y (torch.FloatTensor): Output sequence.
+            x_mask (torch.IntTensor): Input sequence mask.
+            y_mask (torch.IntTensor): Output sequence mask.
+
+        Returns:
+            Tuple[torch.IntTensor, torch.FloatTensor, torch.FloatTensor, torch.FloatTensor]:
+                Durations from the hard alignment map, soft alignment potentials, log scale alignment potentials,
+                hard alignment map.
+
+        Shapes:
+            - x: :math:`[B, T_en, C_en]`
+            - y: :math:`[B, T_de, C_de]`
+            - x_mask: :math:`[B, 1, T_en]`
+            - y_mask: :math:`[B, 1, T_de]`
+
+            - alignment_hard: :math:`[B, T_en]`
+            - alignment_soft: :math:`[B, T_en, T_de]`
+            - alignment_logprob: :math:`[B, 1, T_de, T_en]`
+            - alignment_mas: :math:`[B, T_en, T_de]`
+        """
+        attn_mask = torch.unsqueeze(x_mask, -1) * torch.unsqueeze(y_mask, 2)
+        alignment_soft, alignment_logprob = self.aligner(y.transpose(1, 2), x, x_mask, None)
+        assert not torch.isnan(alignment_soft).any()
+
+        alignment_mas = maximum_path(
+            alignment_soft.squeeze(1).transpose(1, 2).contiguous(), attn_mask.squeeze(1).contiguous()
+        )
+        alignment_hard = torch.sum(alignment_mas, -1).int()
+        alignment_soft = alignment_soft.squeeze(1).transpose(1, 2)
+        return alignment_hard, alignment_soft, alignment_logprob, alignment_mas
+
+    @staticmethod
+    def _set_cond_input(aux_input: Dict):
+        # [TODO] use this
+        return None
+
+    def forward(  # pylint: disable=dangerous-default-value
+        self,
+        tokens: torch.tensor,
+        tokens_lens: torch.tensor,
+        latents: torch.tensor,
+        latents_lengths: torch.tensor,
+        mel: torch.tensor,
+        mel_lens: torch.tensor,
+        pitch: torch.tensor,
+        aux_input={"prompt": None, "durations": None, "language_ids": None},
+    ) -> Dict:
+        outputs = {}
+        tokens_mask = torch.unsqueeze(sequence_mask(tokens_lens, tokens.shape[1]), 1).float()
+        phoneme_enc = self.phoneme_encoder(tokens.unsqueeze(2))
+
+        latent_lens = torch.tensor(latents.shape[1:2]).to(latents.device)
+        mel_mask = torch.unsqueeze(sequence_mask(mel_lens, None), 1).float()
+        latents_mask = torch.unsqueeze(sequence_mask(latent_lens, None), 1).float()
+        # create masks for the segments and remaining parts
+        remaining_mask = torch.ones_like(latents, dtype=torch.bool)
+
+        # Get random segment for the speech prompt
+        speech_prompts, segment_indices = rand_segments(
+            latents, latents_lengths, self.diff_segment_size, let_short_samples=True, pad_short=True
+        )
+
+        # iterate over the batch dimension
+        for i in range(latents.size(0)):
+            remaining_mask[i, :, segment_indices[i] : segment_indices[i] + self.diff_segment_size] = 0
```

Comment:
isn't it easier to just iterate over the `remaning_latents`? `remaining_mask` is not being used somewhere else look like.